typeracer_bot.py

```diff
@@ -1,78 +1,81 @@
+from selenium.common.exceptions import WebDriverException
 from selenium.webdriver.common.by import By
 from selenium.webdriver.support.ui import WebDriverWait
 from selenium.webdriver.support import expected_conditions as EC
 
 
-class TyperacerBot:
-    def __init__(self, link):
-        self.driver = webdriver.Chrome("driver/chromedriver.exe")
-        self.keyboard = Controller()
-        self.link = link
-        self.text = ""
-
-    def start(self):
-        # Open the link
-        self.driver.get(self.link)
-
-    def find_text(self):
-        # Resetting text
-        self.text = ""
-
-        # Get the HTML of the current page
-        src = self.driver.page_source
-        soup = BeautifulSoup(src, "html.parser")
-        span = soup.findAll("span")
-
-        for i in span:
-            if "unselectable" in str(i):
-                self.text += i.text
-
-        if self.text:
-            print("- valid text successfully found")
-
-    def keypress_type(self, delay):
-        for i in self.text:
-            time.sleep(delay)
-            self.keyboard.press(i)
-            self.keyboard.release(i)
-
-    def keysend_type(self, delay):
-        try:
-            # Click the inputbox when it can be clicked
-            wait = WebDriverWait(self.driver, 15)
-            elem = wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "txtInput")))
-        except TimeoutException:
-            print("! inputbox not found within 15s")
-        else:
-            for i in self.text:
-                time.sleep(delay)
-                elem.send_keys(i)
-
-
-def main():
-    # Default link and delay time
-    link = "https://play.typeracer.com"
-    delay = 0.1
-
-    bot = TyperacerBot(link)
-    bot.start()
+def find_text(driver):
+    """Return typeracer text."""
+    # Get the HTML of the current page
+    src = driver.page_source
+    soup = BeautifulSoup(src, "html.parser")
+    span = soup.findAll("span")
+
+    text = ""
+    for i in span:
+        if "unselectable" in str(i):
+            text += i.text
+    return text
+
+def press_text(keyboard, text, delay):
+    """Type `text` using `keyboard` to simulate real keypresses."""
+    for i in text:
+        time.sleep(delay)
+        keyboard.press(i)
+        keyboard.release(i)
+
+def send_text(driver, text, delay):
+    """Send `text` to text input box first working driver in `drivers` dict."""
+    # Click the inputbox when it can be clicked
+    elem = WebDriverWait(driver, 15).until(EC.element_to_be_clickable((By.CLASS_NAME, "txtInput")))
+    for i in text:
+        time.sleep(delay)
+        elem.send_keys(i)
+
+# dict[name: lambda -> WebDriver]
+DRIVER_FUNCS = dict(
+    chrome=lambda: Chrome(executable_path="driver/chromedriver"),
+    firefox=lambda: Firefox(executable_path="driver/geckodriver"),
+    safari=lambda: Safari(executable_path="driver/safaridriver"),
+)
+
+def get_driver(preferred=None, drivers=DRIVER_FUNCS):
```

Comment:
Why not define DRIVER_FUNCS in the function

Reply:
It'll look weird. Plus, it might be useful to outside functions.

Reply:
how about defining it somewhere not in the middle of functions and also vscode gives me this warning `Dangerous default value DRIVER_FUNCS (builtins.dict) as argumentpylint(dangerous-default-value)`

Reply:
lol The warning is because it's a dictionary. I'll move it to just below the imports